### PR TITLE
[FLINK-27847] Support rename/drop column in SchemaManager

### PR DESCRIPTION
**Pull request by zjureel**
`SchemaNamager` supports to rename/drop column in spark

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/schema/SchemaManager.java [hunk 1]
```diff
@@ -191,6 +193,49 @@ public TableSchema commitChanges(List<SchemaChange> changes) throws Exception {
...
+                    }
+                    if (!newFields.removeIf(
+                            f ->
+                                    StringUtils.equalsIgnoreCase(
+                                            f.name(), ((DropColumn) change).fieldName()))) {
+                        throw new IllegalArgumentException(
+                                String.format(
+                                        "The column [%s] doesn't exist in the table[%s].",
+                                        drop.fieldName(), tableRoot));
+                    }
```
Comment (tsreaper) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/schema/SchemaManager.java:238]: Consider a table without partition keys and primary keys. It seems that we can drop all columns of this table, which is obviously undesired. Fix this case and add a test.

File: flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java [hunk 2]
```diff
@@ -281,6 +281,176 @@ public void testAddColumn() throws Exception {
...
+        assertThatThrownBy(
+                        () ->
+                                spark.sql(
+                                        "ALTER TABLE tablestore.default.testRenamePartitionKey RENAME COLUMN a to aa"))
+                .isInstanceOf(RuntimeException.class)
+                .hasMessage("java.lang.UnsupportedOperationException: Cannot rename partition key");
+    }
+
+    @Test
+    public void testDropSingleColumn() throws Exception {
```
Comment (tsreaper) [inline flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java:343]: Test that table can still be read.

File: flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java [hunk 3]
```diff
@@ -281,6 +281,176 @@ public void testAddColumn() throws Exception {
...
+        assertThat(afterRename.toString())
+                .isEqualTo(
+                        "[[CREATE TABLE testDropSingleColumn (\n"
+                                + "  `b` BIGINT,\n"
+                                + "  `c` STRING)\n"
+                                + "]]");
+    }
+
+    @Test
+    public void testDropColumns() throws Exception {
```
Comment (tsreaper) [inline flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java:372]: Test that table can still be read.

File: flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java [hunk 4]
```diff
@@ -281,6 +281,176 @@ public void testAddColumn() throws Exception {
         assertThat(results.toString()).isEqualTo("[[8]]");
     }
 
+    @Test
+    public void testRenameColumn() throws Exception {
```
Comment (tsreaper) [inline flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java:285]: Add these tests:
1. Table can still be read after column names are changed. See `testAddColumn` for how to create such test.
2. Swap two column names
    ```sql
    ALTER TABLE T RENAME COLUMN a to tmp;
    ALTER TABLE T RENAME COLUMN b to a;
    ALTER TABLE T RENAME COLUMN tmp to b;
    ```
    And check that table can still be read and order of columns is not disturbed.

**Review by tsreaper — CHANGES_REQUESTED** [review]

File: flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java (hunk 3 above)
Comment (zjureel) [reply flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java:372]: Read data will fail in this PR. I found this issue [https://issues.apache.org/jira/browse/FLINK-27846](FLINK-27846) before I fix FLINK-27847. Should we fix read data in FLINK-27847 or in FLINK-27846? What do you think of it? HTX

Comment (tsreaper) [reply flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java:372]: "Read data will fail in this PR." What causes this issue? Are there any differences between adding a column and renaming or dropping a column? If there are they should be fixed before merging this PR. That is, let's first implement FLINK-27846 and then come back to this PR.

Comment (zjureel) [reply flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java:372]: > 

Sounds good to me, I will fix [FLINK-27846](https://issues.apache.org/jira/browse/FLINK-27846) first

Comment (JingsongLi) [reply flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java:372]: It's exciting, if [FLINK-27846](https://issues.apache.org/jira/browse/FLINK-27846) can be fixed.

**Comment by JingsongLi** [comment]
Hi @zjureel , we can continue the work of PR~

**Comment by zjureel** [comment]
Hi @JingsongLi @tsreaper I have update the `SchemaManager` and add more test cases in `SchemaEvolutionTest` and `SparkReadITCase` for schema evolution, please help to review them when you're free, THX :)

File: flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java [hunk 5]
```diff
@@ -284,6 +289,204 @@ public void testAddColumn() throws Exception {
...
```
Comment (JingsongLi) [inline flink-table-store-spark/src/test/java/org/apache/flink/table/store/spark/SparkReadITCase.java:293]: Maybe we can create a new class `SparkSchemaChangeITCase`?

**Review by JingsongLi — COMMENTED** [review]
Thanks @zjureel for the update. Looks good to me overall. Just left two minor comments.

**Comment by zjureel** [comment]
Hi @JingsongLi  DONE

**Review by JingsongLi — APPROVED** [review]
Looks good to me!

**Pull request merged**